skada/_subspace.py

```diff
@@ -640,3 +642,322 @@ def TransferJointMatching(
+            raise ValueError("reg should be None or between 0 and 1.")
+        self.reg = 0 if reg is None else reg
+        self.max_iter = max_iter
+        self.tol = tol
+        self.verbose = verbose
+
+    def adapt(self, X, y=None, sample_domain=None, **kwargs):
+        """Predict adaptation (weights, sample or labels).
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            The source data.
+        y : array-like, shape (n_samples,)
+            The source labels.
+        sample_domain : array-like, shape (n_samples,)
+            The domain labels (same as sample_domain).
+
+        Returns
+        -------
+        X_t : array-like, shape (n_samples, n_components)
+            The data transformed to the target subspace.
+        y_t : array-like, shape (n_samples,)
+            The labels (same as y).
+        sample_domain : array-like, shape (n_samples,)
+            The domain labels transformed to the target subspace
+            (same as sample_domain).
+        weights : None
+            No weights are returned here.
+        """
+        X, sample_domain = check_X_domain(
+            X,
+            sample_domain,
+            allow_multi_source=True,
+            allow_multi_target=True,
+        )
+        X_source, X_target = source_target_split(X, sample_domain=sample_domain)
+
+        if X_source.shape[0]:
+            X_source = np.dot(X_source, self.W_)
+        if X_target.shape[0]:
+            X_target = np.dot(X_target, self.W_)
+        # xxx(okachaiev): this could be done through a more high-level API
+        X_adapt, _ = source_target_merge(
+            X_source, X_target, sample_domain=sample_domain
+        )
+        return X_adapt
+
+    def _torch_cov(self, X):
+        """Compute the covariance matrix of X using torch."""
+        torch = self.torch
+        reg = self.reg
+
+        n_samples, d = X.shape
+        X = X - torch.mean(X, dim=0)
+        cov = X.T @ X / n_samples
+        cov = (1 - reg) * cov + reg * torch.trace(cov) * torch.eye(d)
+
+        return cov
+
+    def _D(self, W, X_source, X_target):
+        """Divergence objective function"""
+        torch = self.torch
+
+        Z_source = X_source @ W
+        Z_target = X_target @ W
+
+        sigma_1 = self._torch_cov(Z_source)
+        sigma_2 = self._torch_cov(Z_target)
+        sigma_11 = 2 * sigma_1
+        sigma_12 = sigma_1 + sigma_2
+        sigma_22 = 2 * sigma_2
+
+        L_11 = torch.linalg.cholesky(torch.linalg.inv(sigma_11))
+        L_12 = torch.linalg.cholesky(torch.linalg.inv(sigma_12))
+        L_22 = torch.linalg.cholesky(torch.linalg.inv(sigma_22))
+        Kss = torch.exp(-0.5 * torch.cdist(Z_source @ L_11, Z_source @ L_11))
+        Kst = torch.exp(-0.5 * torch.cdist(Z_source @ L_12, Z_target @ L_12))
+        Ktt = torch.exp(-0.5 * torch.cdist(Z_target @ L_22, Z_target @ L_22))
+
+        return torch.mean(Kss) + torch.mean(Ktt) - 2 * torch.mean(Kst)
```

Comment:
Why don't we return 2 times the value of _D? While it's true that we ultimately multiply it by mu later in the process, omitting this step might lead to confusion for readers accustomed to this method, as they would typically expect to use a specific mu value.

Reply:
There is no factor 2 in equation (8). Did I miss something?

Reply:
No you're right, I mixed up eq 8 and eq 10